src/api/node/sse/client_core.rs

```diff
@@ -0,0 +1,141 @@
+    types::{BoxedEventStream, CoreCommand, EventType, Handler},
+    SseData,
+};
+use eventsource_stream::{Event, Eventsource};
+use futures::stream::TryStreamExt;
+use std::collections::HashMap;
+
+type EventHandler = Box<dyn Fn(SseData) + Send + Sync + 'static>;
+
+pub struct ClientCore {
+    url: String,
+    event_stream: Option<BoxedEventStream>,
+    next_handler_id: u64,
+    event_handlers: HashMap<EventType, HashMap<u64, EventHandler>>,
+    id_types: HashMap<u64, EventType>,
+    is_connected: bool,
+}
+
+impl ClientCore {
+    pub async fn new(url: &str) -> Self {
+        ClientCore {
+            url: url.to_string(),
+            event_stream: None,
+            next_handler_id: 0,
+            event_handlers: HashMap::new(),
+            id_types: HashMap::new(),
+            is_connected: false,
+        }
+    }
+
+    pub async fn connect(&mut self) -> Result<(), ClientError> {
+        // Connect to SSE endpoint.
+        let client = reqwest::Client::new();
+        let response = client.get(&self.url).send().await?;
+
+        let stream = response.bytes_stream();
+        let mut event_stream = stream.eventsource();
+
+        // Handle the handshake with API version.
+        let handshake_event = event_stream
+            .try_next()
+            .await?
+            .ok_or(ClientError::StreamExhausted)?;
+        let handshake_data: SseData = serde_json::from_str(&handshake_event.data)?;
+        let _api_version = match handshake_data {
+            SseData::ApiVersion(v) => Ok(v),
+            _ => Err(ClientError::InvalidHandshake),
+        }?;
+
+        // Wrap stream with box and store it.
+        let boxed_event_stream = Box::pin(event_stream);
+        self.event_stream = Some(boxed_event_stream);
+        self.is_connected = true;
+
+        Ok(())
+    }
+
+    pub fn remove_handler(&mut self, id: u64) -> bool {
+        if let Some(event_type) = self.id_types.get(&id) {
+            match self.event_handlers.get_mut(event_type) {
+                Some(handlers_for_type) => {
+                    self.id_types.remove(&id);
+                    handlers_for_type.remove(&id).is_some()
+                }
+                None => false,
+            }
+        } else {
+            false //not found
+        }
+    }
+
+    pub fn is_connected(&self) -> bool {
+        self.is_connected
+    }
+
+    pub fn handle_event(&mut self, event: Event) -> Result<(), ClientError> {
+        let data: SseData = serde_json::from_str(&event.data)?;
+
+        match data {
+            SseData::ApiVersion(_) => return Err(ClientError::UnexpectedHandshake), // Should only happen once at connection
+            SseData::Shutdown => return Err(ClientError::NodeShutdown),
+
+            // For each type, find and invoke registered handlers
+            event => {
+                if let Some(handlers) = self.event_handlers.get_mut(&event.event_type()) {
+                    for handler in handlers.values() {
```

Comment:
Given this access pattern, perhaps `handlers` could be a `Vec` instead of a `HashMap`?